system_tests/ec2_handler.py

```diff
@@ -0,0 +1,211 @@
+        return self.config['mananger_security_group_name']
+
+
+class EC2Handler(BaseHandler):
+
+    CleanupContext = EC2CleanupContext
+    CloudifyConfigReader = CloudifyEC2InputsConfigReader
+
+    def ec2_client(self):
+        credentials = self._client_credentials()
+        return EC2Connection(**credentials)
+
+    def ec2_infra_state(self):
+
+        ec2_client = self.ec2_client()
+
+        return {
+            'instances': dict(self._instances(ec2_client)),
+            'key_pairs': dict(self._key_pairs(ec2_client)),
+            'elasticips': dict(self._elasticips(ec2_client)),
+            'security_groups': dict(self._security_groups(ec2_client))
+        }
+
+    def ec2_infra_state_delta(self, before, after):
+
+        after = copy.deepcopy(after)
+
+        return {
+            prop: self._remove_keys(after[prop], before[prop].keys())
+            for prop in before.keys()
+        }
+
+    def remove_ec2_resources(self, resources_to_remove):
+
+        ec2_client = self.ec2_client()
+
+        instances = self._instances(ec2_client)
+        key_pairs = self._key_pairs(ec2_client)
+        elasticips = self._elasticips(ec2_client)
+        security_groups = self._security_groups(ec2_client)
+
+        failed = {
+            'instances': {},
+            'key_pairs': {},
+            'elasticips': {},
+            'security_groups': {}
+        }
+
+        for instance_id, _ in instances:
+            if instance_id in resources_to_remove['instances']:
+                with self._handled_exception(instance_id, failed, 'instances'):
+                    ec2_client.terminate_instances(instance_id)
+
+        for kp_name, _ in key_pairs:
+            if kp_name in resources_to_remove['key_pairs']:
+                with self._handled_exception(kp_name, failed, 'key_pairs'):
+                    ec2_client.delete_key_pair(kp_name)
+
+        for elasticip_id, elasticip in elasticips:
+            if elasticip_id in resources_to_remove['elasticips']:
+                with self._handled_exception(
+                        elasticip_id, failed, 'elasticips'):
+                    ec2_client.release_address(elasticip_id)
+
+        for security_group_id, _ in security_groups:
+            if security_group_id in resources_to_remove['security_groups']:
+                with self._handled_exception(
+                        security_group_id, failed, 'security_groups'):
+                    ec2_client.delete_security_group(security_group_id)
+
+        return failed
+
+    def _client_credentials(self):
+
+        return {
+            'aws_access_key_id': self.env.aws_access_key_id,
+            'aws_secret_access_key': self.env.aws_secret_access_key
+        }
+
+    def _security_groups(self, ec2_client):
+        return [(security_group.id, security_group.id)
+                for security_group in ec2_client.get_all_security_groups()]
+
+    def _instances(self, ec2_client):
+        return [(instance[0].id, instance[0].id)
```

Comment:
just making sure there isn't any change in our case that a reservation will contain more than 1 instance